[JSC] Move Loop Unrolling After CPSRethreading to Preserve Variable Availability

hyjorc1 opened the pull request:
#### 8c21b3a3b7cdb94cb0a4424dc158fa710a099093
<pre>
[JSC] Move Loop Unrolling After CPSRethreading to Preserve Variable Availability
<a href="https://rdar.apple.com/149343163">rdar://149343163</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291602">https://bugs.webkit.org/show_bug.cgi?id=291602</a>

Reviewed by Yusuke Suzuki.

In the DFG, loop unrolling was previously performed before CPSRethreading.
While this worked for many optimization cases, it introduced incorrect variable
availability for OSR exit. This is because CPSRethreading is responsible for
recomputing variablesAtHead and variablesAtTail, which are used to determine
which variables are live at the entry and exit of a block.

Performing loop unrolling before CPSRethreading meant that cloned blocks from
the original loop could obscure or disrupt availability information, leading
to missing variables during OSR exit.

This patch moves loop unrolling after CPSRethreading and introduces `m_shouldFixAvailability`
to selectively patch jump pad availability after critical edge breaking,
ensuring correctness in all subsequent compiler phases.

Reviewed by Yusuke Suzuki.

* JSTests/stress/loop-unrolling-liveness-2.js: Added.
(v3.const.v4):
* Source/JavaScriptCore/dfg/DFGBasicBlock.h:
(JSC::DFG::BasicBlock::isJumpPad):
* Source/JavaScriptCore/dfg/DFGCloneHelper.cpp:
(JSC::DFG::CloneHelper::blockClone):
* Source/JavaScriptCore/dfg/DFGCriticalEdgeBreakingPhase.cpp:
(JSC::DFG::CriticalEdgeBreakingPhase::run):
(JSC::DFG::CriticalEdgeBreakingPhase::performFixJumpPadAvailability):
* Source/JavaScriptCore/dfg/DFGCriticalEdgeBreakingPhase.h:
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
(JSC::DFG::LoopUnrollingPhase::LoopData::loopTarget const):
(JSC::DFG::LoopUnrollingPhase::LoopData::exitTarget const):
(JSC::DFG::LoopUnrollingPhase::run):
(JSC::DFG::LoopUnrollingPhase::loopAnalysisPredecessors):
(JSC::DFG::LoopUnrollingPhase::loopAnalysisSuccessor):
(JSC::DFG::LoopUnrollingPhase::locatePreHeader):
(JSC::DFG::LoopUnrollingPhase::locateTail):
(JSC::DFG::LoopUnrollingPhase::unrollLoop):
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::compileInThreadImpl):
* Source/JavaScriptCore/dfg/DFGSSAConversionPhase.cpp:
(JSC::DFG::SSAConversionPhase::run):

Canonical link: <a href="https://commits.webkit.org/293914@main">https://commits.webkit.org/293914@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4be18a3cee6275f6fbba80de2725c929b8fd29b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100273 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19922 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10220 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105403 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50856 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20227 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28396 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76346 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33404 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103280 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15478 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90573 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56703 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15299 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8572 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50224 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/92931 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85209 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8654 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107762 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98880 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27386 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20081 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85315 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27749 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86780 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84837 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29521 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7256 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21279 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16316 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27323 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32556 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/122506 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27134 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34186 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30450 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28693 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGCriticalEdgeBreakingPhase.cpp:
```diff
@@ -95,6 +103,11 @@ bool performCriticalEdgeBreaking(Graph& graph)
+{
+    return CriticalEdgeBreakingPhase::breakCriticalEdge(graph, insertionSet, functor);
+}
+
```
We do not need a change for CriticalEdgeBreaking phase if we just call it later.

hyjorc1 replied:
So, the motivation is that `DFGSSAConversionPhase` needs `variablesAtHead` and `variablesAtTail` for the computation. However, `DFGCriticalEdgeBreakingPhase` only inserts the jump pad without updating the variables since the current FTL optimization flow rely on `performCPSRethreading` which (re)computes the availabilities.

Before:
```
        RUN_PHASE(performCleanUp); // Reduce the graph size a bit.
        RUN_PHASE(performCriticalEdgeBreaking);
        if (Options::createPreHeaders())
            RUN_PHASE(performLoopPreHeaderCreation);
        RUN_PHASE(performCPSRethreading);
        RUN_PHASE(performSSAConversion);
        RUN_PHASE(performSSALowering);
```

Are you suggesting that we can do something like this
```
...
RUN_PHASE(performCPSRethreading);
if (Options::useLoopUnrolling()) {
    RUN_PHASE(performLoopUnrolling);
    RUN_PHASE(performCriticalEdgeBreaking);
    RUN_PHASE(performUpdateAvailabilitiesForJumpPad); // <-- fix the new jump pad with `copyVariablesAsPhis`
}
```

hyjorc1 replied:
Thanks, updated.

Constellation commented on Source/JavaScriptCore/dfg/DFGBasicBlock.h:
```diff
@@ -163,6 +163,34 @@ class BasicBlock {
+                result.append(current);
+        }
+        return result;
+    }
```
I think normal predecessors are anyway real, and `realPredecessors` sounds strange to me.
Let's do,

1. Move this helper to Loop Unrolling Phase's cpp as it is special definition of predecessors only used by that phase.
2. Maybe renaming it to `predecessorsBeforeBreakingCriticalEdges` or something?

hyjorc1 replied:
Fixed

Constellation commented on Source/JavaScriptCore/dfg/DFGBasicBlock.h:
```diff
@@ -163,6 +163,34 @@ class BasicBlock {
+            result = result->successor(0);
+        return result;
+    }
+
```
Ditto. We should move it to loop unrolling phase. And we should rename it to something like `successorBeforeBreakingCriticalEdges` etc.

hyjorc1 replied:
Fixed.

Constellation reviewed (commented):
Some quick comments, still reviewing

Constellation commented on Source/JavaScriptCore/dfg/DFGCriticalEdgeBreakingPhase.cpp:
```diff
@@ -46,6 +46,8 @@ class CriticalEdgeBreakingPhase : public Phase {
     bool run()
     {
+        m_graph.m_newJumpPads.clear();
+
```
Ah, ugh, yeah, this looks a bit weird. So how about doing this?
We first collect newly inserted jump pads here.
```
Vector<BasicBlock*> newJumpPads;
```

Constellation commented on Source/JavaScriptCore/dfg/DFGCriticalEdgeBreakingPhase.cpp:
```diff
@@ -76,6 +78,7 @@ class CriticalEdgeBreakingPhase : public Phase {
+                    m_graph.m_newJumpPads.append(pad);
                 } else
                     pad = iter->value;
 
```
And then, in the end of this phase,
```
bool changed = m_insertionSet.execute();
if (changed && m_graph.m_form == ThreadedCPS)
    fixupJumpPadAvailability(newJumpPads);
return changed;
```

And defining `fixupJumpPadAvailability` as a member of this class.

hyjorc1 replied:
The outputs of `DFGLoopUnrollingPhase.cpp` are `ThreadedCPS` (no unrolling) and `LoadStore` (loop unrolled). The reason is that `BlockInsertionSet::execute()` -> `m_graph.dethread()` which does `m_form = LoadStore;`. I think we can pass a flag via graph after loop unrolling phase in order to trigger `fixupJumpPadAvailability` in  `DFGCriticalEdgeBreakingPhase.cpp`.

Constellation commented on Source/JavaScriptCore/dfg/DFGFixJumpPadAvailabilityPhase.cpp:
```diff
@@ -0,0 +1,93 @@
+bool performFixJumpPadAvailabilityPhase(Graph& graph)
+{
+    return runPhase<FixJumpPadAvailabilityPhase>(graph);
+}
```
So, we will move this to CriticalEdgeBreakingPhase.cpp. We do not need this additional phase.
Since blocks are added in CriticalEdgeBreakingPhase and it breaks ThreadedCPS form, then this fixup should be done in CriticalEdgeBreakingPhase when the form is ThreadedCPS.

hyjorc1 replied:
Removed `performFixJumpPadAvailabilityPhase` and added `performFixJumpPadAvailability` in `DFGCriticalEdgeBreakingPhase.cpp`.

Constellation reviewed (approved):
r=me with changes.

hyjorc1 committed:
[JSC] Move Loop Unrolling After CPSRethreading to Preserve Variable Availability
rdar://149343163
https://bugs.webkit.org/show_bug.cgi?id=291602

Reviewed by Yusuke Suzuki.

In the DFG, loop unrolling was previously performed before CPSRethreading.
While this worked for many optimization cases, it introduced incorrect variable
availability for OSR exit. This is because CPSRethreading is responsible for
recomputing variablesAtHead and variablesAtTail, which are used to determine
which variables are live at the entry and exit of a block.

Performing loop unrolling before CPSRethreading meant that cloned blocks from
the original loop could obscure or disrupt availability information, leading
to missing variables during OSR exit.

This patch moves loop unrolling after CPSRethreading and introduces `m_shouldFixAvailability`
to selectively patch jump pad availability after critical edge breaking,
ensuring correctness in all subsequent compiler phases.

Reviewed by Yusuke Suzuki.

* JSTests/stress/loop-unrolling-liveness-2.js: Added.
(v3.const.v4):
* Source/JavaScriptCore/dfg/DFGBasicBlock.h:
(JSC::DFG::BasicBlock::isJumpPad):
* Source/JavaScriptCore/dfg/DFGCloneHelper.cpp:
(JSC::DFG::CloneHelper::blockClone):
* Source/JavaScriptCore/dfg/DFGCriticalEdgeBreakingPhase.cpp:
(JSC::DFG::CriticalEdgeBreakingPhase::run):
(JSC::DFG::CriticalEdgeBreakingPhase::performFixJumpPadAvailability):
* Source/JavaScriptCore/dfg/DFGCriticalEdgeBreakingPhase.h:
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
(JSC::DFG::LoopUnrollingPhase::LoopData::loopTarget const):
(JSC::DFG::LoopUnrollingPhase::LoopData::exitTarget const):
(JSC::DFG::LoopUnrollingPhase::run):
(JSC::DFG::LoopUnrollingPhase::loopAnalysisPredecessors):
(JSC::DFG::LoopUnrollingPhase::loopAnalysisSuccessor):
(JSC::DFG::LoopUnrollingPhase::locatePreHeader):
(JSC::DFG::LoopUnrollingPhase::locateTail):
(JSC::DFG::LoopUnrollingPhase::unrollLoop):
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::compileInThreadImpl):
* Source/JavaScriptCore/dfg/DFGSSAConversionPhase.cpp:
(JSC::DFG::SSAConversionPhase::run):

Canonical link: https://commits.webkit.org/293914@main
